File: src/notations/number/knuth.c
#include "googology/notations/number/knuth.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

// --- AST node ---
typedef struct knuth_node {
    int is_val;
    goog_int_t val;
    goog_int_t base_val;
    goog_int_t height;
    struct knuth_node* exp;
} knuth_node_t;

static knuth_node_t* knuth_node_value(goog_int_t v) {
    knuth_node_t* n = calloc(1, sizeof(knuth_node_t));
    n->is_val = 1;
    n->val = v;
    return n;
}

static knuth_node_t* knuth_node_arrow(goog_int_t base, goog_int_t h, knuth_node_t* e) {
    knuth_node_t* n = calloc(1, sizeof(knuth_node_t));
    n->is_val = 0;
    n->base_val = base;
    n->height = h;
    n->exp = e;
    return n;
}

static void knuth_node_free(knuth_node_t* n) {
    if (!n) return;
    if (n->exp) knuth_node_free(n->exp);
    free(n);
}

// --- Knuth struct ---
typedef struct {
    goog_notation_t base;
    knuth_node_t* root;
} knuth_t;

// Forward declarations
static const char* knuth_name(const goog_notation_t* n);
static goog_family_t knuth_family(const goog_notation_t* n);
static const char* knuth_creator(const goog_notation_t* n);
static const char* knuth_version(const goog_notation_t* n);
static goog_capabilities_t knuth_capabilities(const goog_notation_t* n);
static void knuth_string_to_it(goog_notation_t* n, const char* s);
static char* knuth_to_latex(const goog_notation_t* n);
static void knuth_expand(goog_notation_t* n, goog_int_t steps);
static void knuth_expand_to(goog_notation_t* n, goog_int_t len);
static int knuth_compare(const goog_notation_t* a, const goog_notation_t* b);
static void knuth_destroy(goog_notation_t* n);

static const goog_notation_vtable_t knuth_vtable = {
    .name = knuth_name,
    .family = knuth_family,
    .creator = knuth_creator,
    .version = knuth_version,
    .capabilities = knuth_capabilities,
    .string_to_it = knuth_string_to_it,
    .to_latex = knuth_to_latex,
    .expand = knuth_expand,
    .expand_to = knuth_expand_to,
    .compare = knuth_compare,
    .destroy = knuth_destroy,
};

// --- Vtable implementations ---
static const char* knuth_name(const goog_notation_t* n) { (void)n; return "knuth"; }
static goog_family_t knuth_family(const goog_notation_t* n) { (void)n; return FAMILY_NUMBER; }
static const char* knuth_creator(const goog_notation_t* n) { (void)n; return "Donald Knuth"; }
static const char* knuth_version(const goog_notation_t* n) { (void)n; return "1"; }

static goog_capabilities_t knuth_capabilities(const goog_notation_t* n) {
    (void)n;
    goog_capabilities_t c = {0};
    goog_capabilities_set(&c, OP_FROM_STRING);
    goog_capabilities_set(&c, OP_TO_STRING);
    goog_capabilities_set(&c, OP_EXPAND);
    goog_capabilities_set(&c, OP_EXPAND_TO);
    return c;
}

// --- Internal helpers ---

static char* tex_node(const knuth_node_t* node) {
    if (node->is_val) {
        char buf[32];
        snprintf(buf, sizeof(buf), "%lld", (long long)node->val);
        return strdup(buf);
    }
    char* bt = tex_node(&(knuth_node_t){.is_val = 1, .val = node->base_val});
    char* et = node->exp ? tex_node(node->exp) : strdup("");
    char arr[64];
    if (node->height == 1)
        snprintf(arr, sizeof(arr), "\\uparrow");
    else
        snprintf(arr, sizeof(arr), "\\uparrow^{%lld}", (long long)node->height);

    int exp_has_arrow = node->exp && !node->exp->is_val;
    char* result;
    if (exp_has_arrow)
        result = malloc(strlen(bt) + strlen(arr) + strlen(et) + 10);
    else
        result = malloc(strlen(bt) + strlen(arr) + strlen(et) + 10);

    if (exp_has_arrow)
        sprintf(result, "%s %s (%s)", bt, arr, et);
    else
        sprintf(result, "%s %s %s", bt, arr, et);

    free(bt); free(et);
    return result;
}

static knuth_node_t* step_node(const knuth_node_t* node) {
    if (node->is_val) {
        return knuth_node_value(node->val);
    }
    if (node->exp && node->exp->is_val) {
        goog_int_t b = node->exp->val;
        if (b == 1) {
            return knuth_node_value(node->base_val);  // a ^c 1 = a
        }
        if (node->height == 1) {
            return knuth_node_arrow(node->base_val, node->height,
                                    knuth_node_value(node->exp->val));
        }
        knuth_node_t* inner = knuth_node_arrow(node->base_val, node->height,
                                                knuth_node_value(b - 1));
        return knuth_node_arrow(node->base_val, node->height - 1, inner);
    }
    // recurse into the rightmost sub-term
    return knuth_node_arrow(node->base_val, node->height, step_node(node->exp));
}

static knuth_node_t* parse_expr(const char** sp) {
    const char* s = *sp;
    while (*s == ' ') s++;

    // Find first '^'
    const char* caret = strchr(s, '^');
    if (!caret) {
        // bare integer
        char* end;
        goog_int_t v = strtoll(s, &end, 10);
        *sp = end;
        return knuth_node_value(v);
    }

    // Parse base
    char base_str[64];
    size_t base_len = (size_t)(caret - s);
    strncpy(base_str, s, base_len);
    base_str[base_len] = '\0';
    goog_int_t base = strtoll(base_str, NULL, 10);

    // Count carets
    int c = 0;
    const char* j = caret;
    while (*j == '^') { c++; j++; }

    // Parse exponent
    *sp = j;
    knuth_node_t* exp = parse_expr(sp);
    return knuth_node_arrow(base, c, exp);
}

static void knuth_string_to_it(goog_notation_t* n, const char* s) {
    knuth_t* k = (knuth_t*)n;
    if (k->root) { knuth_node_free(k->root); k->root = NULL; }

    // Normalize: replace UTF-8 up-arrow with ^, remove spaces
    char* t = strdup(s);
    char* p = t;
    char* q = t;
    while (*p) {
        if ((unsigned char)*p == 0xE2 && (unsigned char)*(p+1) == 0x86 && (unsigned char)*(p+2) == 0x91) {
            *q++ = '^';
            p += 3;
        } else if (!isspace((unsigned char)*p)) {
            *q++ = *p++;
        } else {
            p++;
        }
    }
    *q = '\0';

    if (strlen(t) == 0) { free(t); return; }

    const char* sp = t;
    k->root = parse_expr(&sp);
    free(t);
}

static char* knuth_to_latex(const goog_notation_t* n) {
    const knuth_t* k = (const knuth_t*)n;
    if (!k->root) return strdup("");
    return tex_node(k->root);
}

static void knuth_expand(goog_notation_t* n, goog_int_t steps) {
    knuth_t* k = (knuth_t*)n;
    for (goog_int_t i = 0; i < steps; i++) {
        knuth_node_t* new_root = step_node(k->root);
        knuth_node_free(k->root);
        k->root = new_root;
    }
}

static void knuth_expand_to(goog_notation_t* n, goog_int_t len) {
    char* prev = knuth_to_latex(n);
    while (1) {
        knuth_expand(n, 1);
        char* cur = knuth_to_latex(n);
        if (strlen(cur) >= (size_t)len || strcmp(cur, prev) == 0) {
            free(prev); free(cur);
            return;
        }
        free(prev);
        prev = cur;
    }
}

static int knuth_compare(const goog_notation_t* a, const goog_notation_t* b) {
    (void)a; (void)b;
    return GOOG_ERR_NOT_COMPARABLE;
}

static void knuth_destroy(goog_notation_t* n) {
    knuth_t* k = (knuth_t*)n;
    if (k->root) knuth_node_free(k->root);
    free(k);
}

// --- Public API ---
goog_notation_t* goog_knuth_new(void) {
    knuth_t* k = calloc(1, sizeof(knuth_t));
    k->base.vtable = &knuth_vtable;
    k->root = knuth_node_value(0);
    return (goog_notation_t*)k;
}

goog_notation_t* goog_knuth_parse(const char* s) {
    goog_notation_t* n = goog_knuth_new();
    knuth_string_to_it(n, s);
    return n;
}

File: src/notations/number/knuth.rs
// Knuth up-arrow notation (高德纳箭头).
//   a ^c b = a ^ b                 (c = 1)
//   a ^c b = a                     (b = 1)
//   a ^c b = a ^(c-1) (a ^c (b-1)) (c > 1, b > 1)
//
// Output is LaTeX (e.g. `2 \uparrow\uparrow 3`). The library never computes a
// numeric value: expand()/expand_to()/reduce() rewrite the internal AST and
// return self; to_string() turns it into LaTeX on demand.

use crate::core::capability::{Op, Capabilities};
use crate::core::notation::{Family, Notation, UnsupportedOperation};

/// One node of a Knuth up-arrow expression AST.
#[derive(Clone, Debug)]
enum KNode {
    Value(i64),
    Arrow { base: i64, height: i64, exp: Box<KNode> },
}

impl KNode {
    fn tex(&self) -> String {
        match self {
            KNode::Value(v) => v.to_string(),
            KNode::Arrow { base, height, exp } => {
                let bt = base.to_string();
                let et = exp.tex();
                let arr = if *height == 1 {
                    "\\uparrow".to_string()
                } else {
                    format!("\\uparrow^{{{}}}", height)
                };
                let exp_has_arrow = matches!(exp.as_ref(), KNode::Arrow { .. });
                let exp_part = if exp_has_arrow { format!("({})", et) } else { et };
                format!("{} {} {}", bt, arr, exp_part)
            }
        }
    }

    fn step(&self) -> KNode {
        match self {
            KNode::Value(_) => self.clone(),
            KNode::Arrow { base, height, exp } => {
                match exp.as_ref() {
                    KNode::Value(b) => {
                        if *b == 1 {
                            return KNode::Value(*base); // a ^c 1 = a
                        }
                        if *height == 1 {
                            return self.clone(); // a^b: terminal
                        }
                        let inner = KNode::Arrow {
                            base: *base,
                            height: *height,
                            exp: Box::new(KNode::Value(*b - 1)),
                        };
                        KNode::Arrow {
                            base: *base,
                            height: *height - 1,
                            exp: Box::new(inner),
                        }
                    }
                    _ => {
                        // recurse into the rightmost sub-term
                        KNode::Arrow {
                            base: *base,
                            height: *height,
                            exp: Box::new(exp.step()),
                        }
                    }
                }
            }
        }
    }

    fn parse(s: &str) -> Result<KNode, String> {
        let s = s.trim();
        let pos = s.find('^');
        if pos == None {
            if s.is_empty() {
                return Err("Knuth: empty operand".to_string());
            }
            return Ok(KNode::Value(s.parse::<i64>().map_err(|e| format!("Knuth: {}", e))?));
        }
        let pos = pos.unwrap();
        let a: i64 = s[..pos].parse().map_err(|e| format!("Knuth: {}", e))?;
        let mut j = pos;
        let mut c: i64 = 0;
        while j < s.len() && s.as_bytes()[j] == b'^' {
            c += 1;
            j += 1;
        }
        let rest = s[j..].trim();
        if rest.is_empty() {
            return Err("Knuth: dangling arrows (no exponent)".to_string());
        }
        let exp = KNode::parse(rest)?;
        Ok(KNode::Arrow {
            base: a,
            height: c,
            exp: Box::new(exp),
        })
    }
}

/// Knuth up-arrow notation.
pub struct Knuth {
    root: KNode,
}

impl Knuth {
    pub fn new() -> Self {
        Knuth { root: KNode::Value(0) }
    }

    pub fn from_string(s: &str) -> Result<Self, String> {
        let mut k = Knuth::new();
        k.string_to_it(s)?;
        Ok(k)
    }
}

impl Notation for Knuth {
    fn as_any(&self) -> &dyn std::any::Any { self }
    fn name(&self) -> &str { "knuth" }
    fn family(&self) -> Family { Family::Number }
    fn creator(&self) -> &str { "Donald Knuth" }
    fn version(&self) -> &str { "1" }

    fn capabilities(&self) -> Capabilities {
        let mut c = Capabilities::new();
        c.set(Op::FromString);
        c.set(Op::ToString);
        c.set(Op::Expand);
        c.set(Op::ExpandTo);
        c
    }

    fn string_to_it(&mut self, s: &str) -> Result<(), String> {
        let t = s.replace("↑", "^").replace(' ', "");
        if t.is_empty() {
            return Err("Knuth: empty input".to_string());
        }
        self.root = KNode::parse(&t)?;
        Ok(())
    }

    fn to_string(&self) -> String {
        self.root.tex()
    }

    fn expand(&mut self, n: i64) -> Result<(), String> {
        for _ in 0..n {
            self.root = self.root.step();
        }
        Ok(())
    }

    fn expand_to(&mut self, len: i64) -> Result<(), String> {
        let prev = self.to_string();
        loop {
            self.expand(1)?;
            let cur = self.to_string();
            if cur.len() >= len as usize || cur == prev {
                return Ok(());
            }
            // prev = cur; // would need to rebind
        }
    }
}

impl std::fmt::Display for Knuth {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::core::notation::NotComparable;

    #[test]
    fn test_parse_and_string() {
        let k = Knuth::from_string("2 ^^ 3").unwrap();
        let s = k.to_string();
        println!("  Knuth '2 ^^ 3' -> {s}");
        assert!(s.contains("uparrow"));
    }

    #[test]
    fn test_expand_once() {
        let mut k = Knuth::from_string("2 ^^ 3").unwrap();
        k.expand(1).unwrap();
        let s = k.to_string();
        println!("  Knuth '2 ^^ 3' expand(1) -> {s}");
        assert!(s.contains("uparrow"));
    }

    #[test]
    fn test_not_comparable() {
        let k = Knuth::from_string("2 ^^ 3").unwrap();
        let k2 = Knuth::from_string("3 ^^ 3").unwrap();
        let result = k.compare(&k2);
        assert!(result.is_err());
    }

    #[test]
    fn test_capabilities() {
        let k = Knuth::from_string("2 ^^ 3").unwrap();
        assert!(k.can(Op::FromString));
        assert!(k.can(Op::ToString));
        assert!(k.can(Op::Expand));
        assert!(k.can(Op::ExpandTo));
        assert!(!k.can(Op::Compare));
    }
}